### PR TITLE
Errors through Codes and using BrowserStack for testing

### PR DESCRIPTION
**Pull request by MarcusLongmuir**
This PR removes the `onError` callback in favour of reporting errors through gRPC codes.

It also switches the browser testing to BrowserStack from SauceLabs. This was necessary to test CORS handling.

### CODE REVIEW

File: ts/src/transports/Transport.ts [hunk 1]
```diff
@@ -19,6 +21,31 @@ export type TransportOptions = {
...
 }
 
+let xhr: XMLHttpRequest;
+function getXHR () {
+  if (xhr !== undefined) return xhr;
+
+  if (XMLHttpRequest) {
+    xhr = new XMLHttpRequest();
+    try {
+      xhr.open('GET', 'https://example.com')
```
Comment (mwitkow) [inline ts/src/transports/Transport.ts]: this is weird, why is this needed?

File: ts/src/transports/Transport.ts [hunk 2]
```diff
@@ -33,6 +60,10 @@ export class DefaultTransportFactory {
       return fetchRequest;
     }
 
+    if (xhrSupportsResponseType("ms-stream")) {
```
Comment (mwitkow) [inline ts/src/transports/Transport.ts]: add a comment about this mumbo jumpbo.

File: ts/src/transports/msStream.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import {BrowserHeaders} from "browser-headers";
```
Comment (mwitkow) [inline ts/src/transports/msStream.ts]: Can you add a comment to this whole file as to why this magic is needed + add the doc links to where you found these incantations? :)

File: ts/src/grpc.ts [hunk 4]
```diff
@@ -130,53 +171,62 @@ export namespace grpc {
...
       onChunk: (chunkBytes: Uint8Array) => {
-        props.debug && console.debug("onChunk: ", chunkBytes);
-
         let data: Chunk[] = [];
         try {
           data = parser.parse(chunkBytes);
         } catch (e) {
-          props.onError(e);
+          props.debug && debug("onChunk.parsing error", e, e.message);
+          rawOnError(Code.Internal, e.message);
```
Comment (mwitkow) [inline ts/src/grpc.ts]: Ok, so these are parsing errors. Can you add a "parsing error" prefix to the message?

File: ts/src/grpc.ts [hunk 5]
```diff
@@ -130,53 +171,62 @@ export namespace grpc {
...
       },
       onComplete: () => {
-        props.debug && console.debug("grpc.onComplete");
+        props.debug && debug("grpc.onComplete");
 
         if (responseTrailers === undefined) {
           if (responseHeaders === undefined) {
             // The request was unsuccessful - it did not receive any headers
-            rawOnError(new Error("Response closed without grpc-status (No headers)"));
+            rawOnError(Code.Unknown, "");
```
Comment (mwitkow) [inline ts/src/grpc.ts]: can you add text back to the error?

File: ts/src/grpc.ts [hunk 6]
```diff
@@ -130,53 +171,62 @@ export namespace grpc {
...
+            rawOnMessage(deserialized);
           } else if (d.chunkType === ChunkType.TRAILERS) {
+            props.debug && debug("onChunk.trailers", responseTrailers);
             responseTrailers = new BrowserHeaders(d.trailers);
           }
         });
       },
       onComplete: () => {
-        props.debug && console.debug("grpc.onComplete");
+        props.debug && debug("grpc.onComplete");
```
Comment (mwitkow) [inline ts/src/grpc.ts]: what happens on resolution failures etc? is that gonna go through onComplete? If so, can you make sure we return a proper error here? and maybe add a comment as to why :)

File: ts/src/transports/Transport.ts (hunk 1 above)
Comment (MarcusLongmuir) [inline ts/src/transports/Transport.ts]: This can be `https://localhost`. The transport feature detection requires an `XMLHttpRequest`, which can only be constructed with a valid URL. It doesn't actually make a request.

File: ts/src/transports/Transport.ts (hunk 2 above)
Comment (MarcusLongmuir) [inline ts/src/transports/Transport.ts]: Will add in the `ms-stream.ts` file

**Comment by mwitkow** [comment]
LGTM 🚢
